runtime/storage/src/trie/mod.rs

```diff
@@ -1041,4 +1437,78 @@ mod tests {
+
+    fn gen_changes(rng: &mut ThreadRng) -> Vec<(Vec<u8>, Option<Vec<u8>>)> {
+        let alphabet = &b"abcdefgh"[0..rng.gen_range(2, 8)];
+        let max_length = rng.gen_range(2, 8);
+
+        let mut state: HashMap<Vec<u8>, Vec<u8>> = HashMap::new();
+        let mut result = Vec::new();
+        let delete_probability = rng.gen_range(0.1, 0.5);
+        let size = rng.gen_range(1, 20);
+        for _ in 0..size {
+            let key_length = rng.gen_range(1, max_length);
+            let key: Vec<u8> =
+                (0..key_length).map(|_| alphabet.choose(rng).unwrap().clone()).collect();
+
+            let delete = rng.gen_range(0.0, 1.0) < delete_probability;
+            if delete {
+                let mut keys: Vec<_> = state.keys().cloned().collect();
+                keys.push(key);
+                let key = keys.choose(rng).unwrap().clone();
+                state.remove(&key);
+                result.push((key.clone(), None));
+            } else {
+                let value_length = rng.gen_range(1, max_length);
+                let value: Vec<u8> =
+                    (0..value_length).map(|_| alphabet.choose(rng).unwrap().clone()).collect();
+                result.push((key.clone(), Some(value.clone())));
+                state.insert(key, value);
+            }
+        }
+        result
+    }
+
+    fn simplify_changes(
+        changes: &Vec<(Vec<u8>, Option<Vec<u8>>)>,
+    ) -> Vec<(Vec<u8>, Option<Vec<u8>>)> {
+        let mut state: HashMap<Vec<u8>, Vec<u8>> = HashMap::new();
+        for (key, value) in changes.iter() {
+            if let Some(value) = value {
+                state.insert(key.clone(), value.clone());
+            } else {
+                state.remove(key);
+            }
+        }
+        let mut result: Vec<_> = state.into_iter().map(|(k, v)| (k, Some(v))).collect();
+        result.sort();
+        result
+    }
+
+    #[test]
+    fn test_trie_unique() {
+        let mut rng = rand::thread_rng();
+        for _ in 0..100 {
+            let trie = create_trie();
+            let trie_changes = gen_changes(&mut rng);
+            let simplified_changes = simplify_changes(&trie_changes);
+
+            let (db_changes1, root1) =
+                trie.update(&Trie::empty_root(), trie_changes.iter().cloned());
+            let (db_changes2, root2) =
+                trie.update(&Trie::empty_root(), simplified_changes.iter().cloned());
+            if root1 != root2 {
+                eprintln!("{:?}", trie_changes);
+                eprintln!("{:?}", simplified_changes);
+                eprintln!("root1: {}", root1);
+                eprintln!("root2: {}", root2);
+                panic!("MISMATCH!");
+            }
+            let mut db_changes1: Vec<_> = db_changes1.into_iter().collect();
+            db_changes1.sort();
+            let mut db_changes2: Vec<_> = db_changes2.into_iter().collect();
+            db_changes2.sort();
+            assert_eq!(db_changes1, db_changes2);
```

Comment:
I'm not 100% sure what are you trying to test here - I thought you are already internally sorting changes sets, should then they match even without sorting before?

Reply:
DBChanges is a hashmap, will probably match anyway because we insert in the same order